fix: implement IdentifiableBpmnElement with template pattern for ID v…

giampa91 opened the pull request:
…alidation

Introduces IdentifiableBpmnElement interface for BPMN elements with identifiers. 

## Description

<!-- Describe the goal and purpose of this PR. -->

Implements a custom validator using the template pattern to systematically validate BPMN elements missing required IDs.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)
  - [ ] enable backports [when recommended](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)

## Related issues

closes #26248

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/camunda/camunda?pullRequest=31429) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/camunda/camunda?pullRequest=31429) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/camunda/camunda?pullRequest=31429) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/camunda/camunda?pullRequest=31429) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/camunda/camunda?pullRequest=31429) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/camunda/camunda?pullRequest=31429) it.</sub>

korthout commented on zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/ActivityValidator.java:
```diff
@@ -17,10 +17,10 @@
 
-public class ActivityValidator implements ModelElementValidator<Activity> {
+public class ActivityValidator extends IdentifiableBpmnElementValidator<Activity> {
+
```
❌ Let's [avoid inheritance as a way to reuse code](https://github.com/camunda/camunda/wiki/Code-Style#favor-final-modifier). Please use composition instead: Call the `validate` method of `IdentifiableBpmnElementValidator` inside this class's `validate` method, either by making `IdentifiableBpmnElementValidator#validate` a static method, or keep an instance of `IdentifiableBpmnElementValidator` as a field and call the instance method.

giampa91 replied:
Thanks korthout, I decided to use a static method, because in the end we don't use a complex validation for id that require a dependency.

korthout commented on zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/ValidationVisitor.java:
```diff
@@ -18,6 +18,7 @@
 import static java.util.stream.Collectors.groupingBy;
 
 import io.camunda.zeebe.model.bpmn.instance.BpmnModelElementInstance;
+import io.camunda.zeebe.model.bpmn.instance.TimerEventDefinition;
```
❌ Please remove this redundant import

korthout reviewed (changes requested):
Thanks @giampa91 for your first contribution to this repo! 🙌 

💭 Apologies for having to wait so long before our reply.

👍 I really like the idea to move the `id` to a separate reusable interface.
👍 The change in the api is backward compatible.

❌ We favor composition over inheritance, especially when concerning code reuse. Please adjust the validation changes accordingly.

korthout reviewed (changes requested):
Thanks a lot @giampa91 🚀 

:x: I have one last request before we can merge this:
- Please add a test case to [`ValidateProcessTest`](https://github.com/camunda/camunda/blob/main/zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/validation/ValidateProcessTest.java) that shows that we require process to have an ID
- Bonus: Test it for all IdentifiableBpmnElements

🔧 Please also rebase your changes [rather than using a merge commit to](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#updating-a-pull-request).

korthout commented on zeebe/bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/IdentifiableBpmnElementValidator.java:
```diff
@@ -0,0 +1,29 @@
+  public static <T extends IdentifiableBpmnElement> void validate(
+      final T element, final ValidationResultCollector validationResultCollector) {
+    if (element.getId() == null || element.getId().isEmpty()) {
+      validationResultCollector.addError(0, "Link id must be present and not empty.");
```
❌ I just noticed this when you added the test case.

```suggestion
      validationResultCollector.addError(0, "Element id must be present and not empty.");
```

korthout reviewed (changes requested):
:x: your changes are failing the [Reject merge commits](https://github.com/camunda/camunda/actions/runs/14945996057/job/42145668702) CI workflow. Please rebase your changes [rather than using a merge commit to](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#updating-a-pull-request).

giampa91 commented:
Thank you @korthout  🚀

- I've adjusted the commit message to comply with commitlint. 
- Additionally, I've reviewed the failing tests and removed the ID validation from the ZeebeLoopCharacteristicsValidator and other classes where it wasn't actually required.
- I've also squashed all the changes into a single commit for a cleaner history. 
- Finally, I ran mvn spotless:apply again just to be sure everything is formatted correctly.

Hopefully, the CI pipeline should pass this time, if I am not missing something else. 
Let me know if there's anything else I can do!

korthout reviewed (changes requested):
I had another look to ensure that we don't unnecessarily break backward compatibility with process models that could be deployed previously.

korthout reviewed (approved):
🙇 I've added a commit to remove the unnecessary validations.

With that, I think we can merge this 👍 LGTM 

🚀 Thanks again for this great contribution @giampa91
